Modules/mathmodule.c

```diff
@@ -2561,8 +2561,8 @@ math_prod_impl(PyObject *module, PyObject *iterable, PyObject *start)
                 long x = i_result * b;
                 /* Continue if there is no overflow */
                 if (overflow == 0
-                    && x < LONG_MAX && x > LONG_MIN
-                    && !(b != 0 && x / b != i_result)) {
+                    && !(b != 0 && (i_result > LONG_MAX / b
+                                || i_result < LONG_MIN / b))) {
```

Comment:
There's a nasty corner case here: `LONG_MIN / b` gives undefined behaviour on a typical two's complement machine when `b = -1`.

Reply:
Should we adopt a similar strategy as you pointed out inhttps://github.com/python/cpython/blob/2f1a317d5fdc45b9d714b067906f612f636ba08e/Objects/intobject.c#L496-L518 ?

Reply:
That's one possible solution; it's not my favourite, but it has the advantage of being well battle-tested.

The Python 2.7 int * int overflow detection needs to be a little bit stricter than what you need here, though, in that here it's okay if you have the occasional false positive (overflow detected even though we didn't actually strictly overflow), while in the Python 2.7 code the overflow detection needs to be exact.

So you might be able to get away with something weaker.

Reply:
Should we just avoid the second check if `b == -1`?